score.go

```diff
@@ -153,12 +149,56 @@ func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands [
 	// Use standard parameter defaults (used in Lucene and academic papers)
 	k, b := 1.2, 0.75
 	sumTf := 0.0 // Just for debugging
-	score := 0.0
-	for _, freq := range termFreqs {
+
+	tfs := make(termFrequencyScore)
+
+	for term, freq := range termFreqs {
 		tf := float64(freq)
 		sumTf += tf
-		score += ((k + 1.0) * tf) / (k*(1.0-b+b*L) + tf)
+
+		// Invariant: the keys of df are the union of the keys of tfs over all files.
+		df[term] += 1
+		tfs[term] = ((k + 1.0) * tf) / (k*(1.0-b+b*L) + tf)
 	}
 
-	fileMatch.addBM25Score(score, sumTf, L, opts.DebugScore)
+	if opts.DebugScore {
+		fileMatch.Debug = fmt.Sprintf("sum-termFrequencyScore: %.2f, length-ratio: %.2f", sumTf, L)
+	}
+
+	return tfs
+}
+
+// idf computes the inverse document frequency for a term. nq is the number of
+// documents that contain the term and documentCount is the total number of
+// documents in the corpus.
+func idf(nq, documentCount int) float64 {
+	return math.Log(1.0 + ((float64(documentCount) - float64(nq) + 0.5) / (float64(nq) + 0.5)))
+}
+
+// termDocumentFrequency is a map "term" -> "number of documents that contain the term"
+type termDocumentFrequency map[string]int
+
+// termFrequencyScore is a map "term" -> "term frequency score"
+type termFrequencyScore map[string]float64
+
+// fileMatchesWithScores is a helper type that is used to store the file matches
+// along with internal scoring information.
+type fileMatchesWithScores struct {
+	fileMatches []FileMatch
+	tfs         []termFrequencyScore
+}
+
+func (m *fileMatchesWithScores) addFileMatch(fm FileMatch, tfs termFrequencyScore) {
+	m.fileMatches = append(m.fileMatches, fm)
+	m.tfs = append(m.tfs, tfs)
+}
+
+func (m *fileMatchesWithScores) scoreFilesUsingBM25(df termDocumentFrequency, documentCount int) {
```

Comment:
As a small tweak, we could collect the term frequencies instead of scores, then perform the full BM25 calculation in `scoreFilesUsingBM25`. I like that because it keeps the BM25 calculation in one place, which makes it easier to document and understand the formula. Also we could only have one data type that's `map[string]int` (maybe renaming `termDocumentFrequency` -> `frequencies`) ?

Also we could move the `fileMatch.Debug` calculation to this method to report both IDF and the term frequency component?